vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java

```diff
@@ -88,23 +94,57 @@
 
     private SerializableConsumer<UI> sizeRequest;
 
+    private CheckboxGroupDataView<T> dataView;
+
     public CheckboxGroup() {
         super(Collections.emptySet(), Collections.emptySet(), JsonArray.class,
                 CheckboxGroup::presentationToModel,
                 CheckboxGroup::modelToPresentation);
         registerValidation();
     }
 
+    @Override
+    public CheckboxGroupDataView<T> setDataSource(
+            DataProvider<T, ?> dataProvider) {
+        setDataProvider(dataProvider);
+        return getDataView();
+    }
 
     @Override
-    public CheckboxGroupListDataView<T> setDataProvider(ListDataProvider<T> dataProvider) {
-        this.setDataProvider((DataProvider<T, ?>) dataProvider);
+    public CheckboxGroupListDataView<T> setDataSource(
+            ListDataProvider<T> dataProvider) {
+        this.setDataProvider(dataProvider);
         return getListDataView();
     }
 
     @Override
     public CheckboxGroupListDataView<T> getListDataView() {
-        return new CheckboxGroupListDataView<>(this::getDataProvider, this);
+        if (getDataProvider() instanceof ListDataProvider) {
+            if (dataView == null || !(dataView instanceof ListDataView)) {
+                dataView = new CheckboxGroupListDataView<>(
+                        this::getDataProvider, this);
+            }
+            return (CheckboxGroupListDataView) dataView;
+        }
+        throw new IllegalStateException(
+                "Required ListDataProvider, but got " + getDataProvider()
```

Comment:
It was stateful at some point. The main reason for storage is now that if you create a LazyDataView you should get the LazyDataView from getDataView also and not a recreated generic one. We can ofc if-else the generic one.

Instead of the thrown exception we can just have it die on the dataView verification.